### PR TITLE
Add datasaver_builder and dond_core

### PR DESCRIPTION
**Pull request by samantha-ho**
datasaver_builder and dond_core are intermediate abstractions designed to be used between the low level Measurement context and the high level doNd functions. They allow for greater flexibility in measurement than doNd without requiring parameter registration and explicit datasaver calls.

### CODE REVIEW

**Comment by samantha-ho** [comment]
Last thing to do is replace the example with an actual tutorial on using the datasaver_builder.

**Review by astafan8 — COMMENTED** [review]
(sorry, apparently i reviewed last week but didn't submit these comments; apologies if some of them are obsolete; i'll do another review of the latest changes tomorrow)

File: qcodes/dataset/measurement_extensions.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    measurement_instances = setup_measurement_instances(dataset_definitions, experiment)
+    with catch_interrupts() as interrupted, ExitStack() as stack:
+        datasavers = [
+            stack.enter_context(measurement.run())
+            for measurement in measurement_instances
+        ]
+        try:
+            yield datasavers
+        except Exception:
+            raise Exception
```
[the remark this answers is not in the document]
Comment (samantha-ho) [reply qcodes/dataset/measurement_extensions.py]: The intention was to re-raise any thrown exceptions for the nested context managers, but maybe removing this entirely is appropriate.

File: qcodes/dataset/measurement_extensions.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            if set_event.get_after_set:
+                results[set_event.parameter] = set_event.parameter()
+            else:
+                results[set_event.parameter] = set_event.new_value
+
+        meas_value_pair = process_params_meas(params_meas)
+        for meas_param, value in meas_value_pair:
+            results[meas_param] = value
+
+        filtered_results_list = [(param, value) for param, value in results.items()]
```
[the remark this answers is not in the document]
Comment (samantha-ho) [reply qcodes/dataset/measurement_extensions.py]: This has been now been changed.

File: qcodes/dataset/measurement_extensions.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            meas.register_parameter(param)
+        for param in ds_def.dependent:
+            meas.register_parameter(param, setpoints=ds_def.independent)
+        measurements.append(meas)
+    return measurements
+
+
+@contextmanager
+def datasaver_builder(
+    dataset_definitions: Sequence[DataSetDefinition], experiment: Experiment
```
[the remark this answers is not in the document]
Comment (samantha-ho) [reply qcodes/dataset/measurement_extensions.py:58]: I'd missed that detail. Thanks for the suggestion.

File: qcodes/tests/dataset/test_measurement_extensions.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+
+@pytest.fixture
+def default_database_and_experiment(tmp_path):
+    db_path = Path(tmp_path) / "context_tests.db"
+    qc.initialise_or_create_database_at(db_path)
+    experiment = qc.load_or_create_experiment("context_tests")
+    yield experiment
+    conn = connect(db_path)
+    conn.close()
```
[the remark this answers is not in the document]
Comment (samantha-ho) [reply qcodes/tests/dataset/test_measurement_extensions.py:91]: In my experience, deleting the experiment instance is not always sufficient to remove the wal and shm files. conn.close() has been my go-to, once all other references to the database are removed.

File: qcodes/dataset/measurement_extensions.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            results[meas_param] = value
+
+        filtered_results_list = [(param, value) for param, value in results.items()]
+        datasaver.add_result(
+            *filtered_results_list,
+            *additional_setpoints_data,
+        )
+
+
+class LinSweeper(LinSweep):
```
[the remark this answers is not in the document]
Comment (samantha-ho) [reply qcodes/dataset/measurement_extensions.py:168]: Unsure. I think this is a feature which could be incorporated directly into the AbstractSweep class, but I'd want to think carefully about how to do that.

Also, if this proves useful, I could imagine wanting more of these "AbstractSweeper" objects. For now, I think LinSweeper covers most cases, and we could an ArraySweeper in a later PR.

File: docs/examples/DataSet/Datasaver_Builder.ipynb
Comment (astafan8) [inline docs/examples/DataSet/Datasaver_Builder.ipynb:1]: For the documentation to render correctly, please use the levels of the headers: so e.g. "Setup" section should have level 2, so ``##``, as oppposed to just 1 ``#``, Etc.

**Review by astafan8 — APPROVED** [review]

**Pull request merged**